runners/core-construction-java/src/main/java/org/apache/beam/runners/core/construction/graph/QueryablePipeline.java

```diff
@@ -0,0 +1,277 @@
+            network.successors(producedNode));
+        unproducedCollections.remove(producedNode);
+      }
+      for (Map.Entry<String, String> consumed : transform.getInputsMap().entrySet()) {
+        // This loop may add an edge between the consumed PCollection and the current PTransform.
+        // The local name of the transform must be used to determine the type of edge.
+        PCollectionNode consumedNode = pCollectionNode(consumed.getValue());
+        if (network.addNode(consumedNode)) {
+          // This node has been added to the network for the first time, so it has no producer.
+          unproducedCollections.add(consumedNode);
+        }
+        if (getLocalSideInputNames(transform).contains(consumed.getKey())) {
+          network.addEdge(consumedNode, transformNode, new SingletonEdge());
+        } else {
+          network.addEdge(consumedNode, transformNode, new PerElementEdge());
+        }
+      }
+    }
+    checkState(
+        unproducedCollections.isEmpty(),
+        "%ss %s were consumed but never produced",
+        PCollectionNode.class.getSimpleName(),
+        unproducedCollections);
+    return network;
+  }
+
+  /**
+   * Get the transforms that are roots of this {@link QueryablePipeline}. These are all nodes which
+   * have no input {@link PCollection}.
+   */
+  public Set<PTransformNode> getRootTransforms() {
+    return pipelineNetwork
+        .nodes()
+        .stream()
+        .filter(pipelineNode -> pipelineNetwork.inEdges(pipelineNode).isEmpty())
+        .map(pipelineNode ->  (PTransformNode) pipelineNode)
+        .collect(Collectors.toSet());
+  }
+
+  public PTransformNode getProducer(PCollectionNode pcollection) {
+    return (PTransformNode) Iterables.getOnlyElement(pipelineNetwork.predecessors(pcollection));
+  }
+
+  /**
+   * Get all of the {@link PTransformNode PTransforms} which consume the provided {@link
+   * PCollectionNode} on a per-element basis.
+   *
+   * <p>If a {@link PTransformNode} consumes a {@link PCollectionNode} on a per-element basis one or
+   * more times, it will appear a single time in the result.
+   *
+   * <p>In theory, a transform may consume a single {@link PCollectionNode} in both a per-element
+   * and singleton manner. If this is the case, the transform node is included in the result, as it
+   * does consume the {@link PCollectionNode} on a per-element basis.
+   */
+  public Set<PTransformNode> getPerElementConsumers(PCollectionNode pCollection) {
+    return pipelineNetwork
+        .successors(pCollection)
+        .stream()
+        .filter(
+            consumer ->
+                pipelineNetwork
+                    .edgesConnecting(pCollection, consumer)
+                    .stream()
+                    .anyMatch(PipelineEdge::isPerElement))
+        .map(pipelineNode -> (PTransformNode) pipelineNode)
+        .collect(Collectors.toSet());
+  }
+
+  public Set<PCollectionNode> getProducedPCollections(PTransformNode ptransform) {
```

Comment:
nit: consider using getOutputPCollections since it mirrors the terms output map and input map on the protos.

Reply:
Done.